### DISCUSSION TITLE
UserAgent: memory leak

### DISCUSSION THREAD

**Discussion by dpanech** [mojolicious/mojo#2076] [Q&A]
* Mojolicious version: 9.31
* Perl version: v5.16.3
* Operating system: CentOS 7

### Steps to reproduce the behavior
UserAgent leaks, here's a minimal example:

```perl
#!/usr/bin/perl

use warnings;
use strict;
use Mojo::IOLoop;
use Mojo::UserAgent;

# IOLoop
my $IOLOOP = Mojo::IOLoop->singleton;

# Create UA
my $UA =  Mojo::UserAgent->new;
$UA->ioloop ($IOLOOP);  # not sure if this is necessary

# Make a client request
my $COUNT = 0;
sub ua_test_1($$) {
    my ($host, $port) = @_;

    my $url = "http://$host:$port/";
    ++$COUNT;
    $UA->get_p ($url)
        ->catch (sub {
            print STDERR $host, ':', $port, ': ', @_, " [COUNT=$COUNT]\n"
        })
        ->finally (sub {
            $IOLOOP->next_tick (sub { ua_test_1 ($host, $port) });
            --$COUNT;
        });
}

ua_test_1 ("localhost", 1234);  # some unused port for "connection refused"

print "PID=$$\n";
print "Press ENTER to start\n";
readline();
$IOLOOP->start();
```

### Expected behavior
Memory usage should remain steady while this script is running.

### Actual behavior
Memory reported by top (the `VIRT` column) steadily increases by approx 20 KiB per second.

Note that the example above doesn't involve self-referencing closures - AFAICT - so it can't be that.

**Comment by kraih** [comment]
Not a leak, you're just not giving the event loop any time to do actual work.

**Comment by dpanech** [comment]
Sorry, what work? I'm scheduling next tick after the request has completed, no ?

**Comment by dpanech** [comment]
I just tried replacing next_tick with

```perl
$IOLOOP->timer (0.01 => sub { ua_test_1 ($host, $port) });
```

Memory usage still increases, but more slowly.

**Reply by dpanech** [reply]
Same with `timer (0.1 => ...`

**Comment by dpanech** [comment]
Upgraded to perl 5.24 and it's not happening anymore!

**Reply by kraih** [reply]
Interesting, i've only tested it on Perl 5.36.
